src/Domain/Traits/ScrubByFamily.php

```diff
@@ -0,0 +1,68 @@
+<?php
+/*
+Gibbon, Flexible & Open School System
+Copyright (C) 2010, Ross Parker
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+namespace Gibbon\Domain\Traits;
+
+/**
+ * Implements the ScrubbableGateway interface.
+ */
+trait ScrubByFamily
+{
+    public function getScrubbableRecords(string $cutoffDate, array $context = []) : array
+    {
+        $tableName = $this->getTableName();
+        $scrubbableKey = $this->getScrubbableKey();
+
+        $query = $this
+            ->newSelect()
+            ->cols([$this->getTableName().'.'.$this->getPrimaryKey(), 'gibbonPerson.gibbonPersonID'])
+            ->from($this->getTableName())
+            ->where("gibbonPerson.status='Left'");
+
+        // Join the correct family relation based on the role category
+        if (in_array('Student', $context)) {
+            $query->innerJoin('gibbonFamilyChild', "gibbonFamilyChild.gibbonFamilyID={$tableName}.{$scrubbableKey}")
+                ->innerJoin('gibbonPerson', 'gibbonPerson.gibbonPersonID=gibbonFamilyChild.gibbonPersonID');
+        } else {
+            $query->innerJoin('gibbonFamilyAdult', "gibbonFamilyAdult.gibbonFamilyID={$tableName}.{$scrubbableKey}")
+                ->innerJoin('gibbonPerson', 'gibbonPerson.gibbonPersonID=gibbonFamilyAdult.gibbonPersonID');
+        }
+
+        // Apply the context based on user role
+        if (!empty($context)) {
+            $query->innerJoin('gibbonRole', 'gibbonRole.gibbonRoleID=gibbonPerson.gibbonRoleIDPrimary')
+                ->where('FIND_IN_SET(gibbonRole.category, :roleCategories)')
+                ->bindValue('roleCategories', implode(',', $context));
+        }
+
+        // Only get users whose dateEnd is before the cutoff, falling back to using the lastTimestamp
+        $query->where('((gibbonPerson.dateEnd IS NOT NULL AND gibbonPerson.dateEnd < :cutoffDate) 
+            OR (gibbonPerson.lastTimestamp IS NOT NULL AND gibbonPerson.lastTimestamp < :cutoffDate) 
```

Comment:
Should this line also contain a gibbonPerson.dateEnd IS NOT NULL, to ensure it only looks at last login if the end date is not set? I guess in theory there should not be a last login date after the end date, unless the end date was retroactively entered.

Reply:
Sure, I'll add that to be on the safe side.